### PR TITLE
Treewide: change all occurrences of "name" to "pname-version"

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
Related issue: https://github.com/NixOS/nixpkgs/issues/103997

**Comment by siraben** [comment]
Is it expected that none of these derivations will build differently because the interpolated strings are the same as they were before?

**Comment by AndersonTorres** [comment]
Yes. It is just a rewriting of the expressions.

File: pkgs/applications/audio/google-musicmanager/default.nix [hunk 1]
```diff
@@ -14,7 +14,7 @@ stdenv.mkDerivation rec {
   # which will contain the links to all available *.debs for the arch.
 
   src = fetchurl {
-    url    = "http://dl.google.com/linux/musicmanager/deb/pool/main/g/google-musicmanager-beta/${name}_amd64.deb";
+    url    = "http://dl.google.com/linux/musicmanager/deb/pool/main/g/google-musicmanager-beta/${pname}-${version}_amd64.deb";
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/google-musicmanager/default.nix:17]: Unrelated to this PR but the platform should probably be set to x84_64 only.

File: pkgs/applications/audio/jackmix/default.nix [hunk 2]
```diff
@@ -1,9 +1,11 @@
 { stdenv, fetchurl, pkgconfig, sconsPackages, qt4, lash, libjack2, jack ? libjack2, alsaLib }:
 
 stdenv.mkDerivation {
-  name = "jackmix-0.5.2";
+  pname = "jackmix";
+  version = "0.5.2";
+
   src = fetchurl {
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/jackmix/default.nix:7]: This should probably use fetchFromGitHub.

File: pkgs/applications/audio/mp3gain/default.nix [hunk 3]
```diff
@@ -1,9 +1,11 @@
 { stdenv, fetchurl, unzip, mpg123 }:
 
 stdenv.mkDerivation {
-  name = "mp3gain-1.6.2";
+  pname = "mp3gain";
+  version = "1.6.2";
+
   src = fetchurl {
-    url = "mirror://sourceforge/mp3gain/mp3gain-1_6_2-src.zip";
+    url = "mirror://sourceforge/mp3gain/${pname}-1_6_2-src.zip";
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/mp3gain/default.nix:8]:
```suggestion
    url = "mirror://sourceforge/mp3gain/${pname}-${builtins.replaceStrings ["."] ["_"] version}-src.zip";
```

I did not test this.

File: pkgs/applications/audio/jackmix/default.nix (hunk 2 above)
Comment (AndersonTorres) [reply pkgs/applications/audio/jackmix/default.nix:7]: It would change the final output. I am not trying to do this here and now.

**Review by Atemu — CHANGES_REQUESTED** [review]
Doesn't eval for the most part.

Fix: https://github.com/AndersonTorres/nixpkgs/pull/1

**Comment by Atemu** [comment]
With my changes, Nixpkgs can be eval'd again and the sources that aren't already broken still produce the same output.

Also, nit: Treewide -> treewide

**Comment by AndersonTorres** [comment]
Oh man, this will not be easy.

There are many "legitimate" uses of `name`in Nixpkgs, in the sense that merely changing a single file from `name` to `pname-version` triggers auxiliary functions defined elsewhere - and these functions use `name`!

File: pkgs/applications/audio/mp3gain/default.nix (hunk 3 above)
Comment (doronbehar) [reply pkgs/applications/audio/mp3gain/default.nix]: This is also out of scope for this PR.

**Comment by doronbehar** [comment]
I think it's impossible for this to not trigger rebuilds since even in the simplest case, the builder gets after this change the new environment variable `pname` and `version` which weren't there before and unfortunately (without structuredAttrs I suppose) it's a change in the inputs.

**Comment by AndersonTorres** [comment]
@doronbehar : then, in a certain sense, the expression itself becomes a "build input" of itself?

And what is structuredAttrs?

Well, in this sense I think we can sprint it to the next viable update.

**Comment by doronbehar** [comment]
> And what is structuredAttrs?

https://github.com/NixOS/nixpkgs/pull/72074 && https://nixos.mayflower.consulting/blog/2020/01/20/structured-attrs/

> @doronbehar : then, in a certain sense, the expression itself becomes a "build input" of itself?

I'm not sure I understand your explanation, but to demonstrate how easily rebuilds can be triggered, consider a change such as:

```diff
diff --git i/pkgs/development/compilers/gcc/9/default.nix w/pkgs/development/compilers/gcc/9/default.nix
index 305ed56df78..e39c787284b 100644
--- i/pkgs/development/compilers/gcc/9/default.nix
+++ w/pkgs/development/compilers/gcc/9/default.nix
@@ -83,6 +83,7 @@ in
 stdenv.mkDerivation ({
   pname = "${crossNameAddon}${name}${if stripped then "" else "-debug"}";
   inherit version;
+  dummyVar = "bla";
 
   builder = ../builder.sh;
 
```

That would rebuild the world, and you can see that via running `nix show-derivation -f. gcc.cc` before and after the change, and inspect the diff in a text editor.

The difference between the derivations can be found in the `env` attribute of the derivations, where the `dummyVar` environment variable which you and I know that won't affect the build is the culprint, but Nix doesn't know it.

In general `nix show-derivation -f.` is the judge, though ofborg has a more sophisticated way of doing that in a batch.

What's so frustrating about it is that our `default-builder.sh` and most of it's derivatives handle the conditionals which ideally should have been evaluated by Nix. A real world examples of such unnecessary rebuilds triggered is (only the Darwin rebuilds of) e.g: https://github.com/NixOS/nixpkgs/issues/94952#issuecomment-673557750 .

Even changes in the **order** of inputs in a Nix arrays causes rebuilds, consider this change:

```diff
diff --git i/pkgs/development/compilers/gcc/9/default.nix w/pkgs/development/compilers/gcc/9/default.nix
index 305ed56df78..d73089e9c0c 100644
--- i/pkgs/development/compilers/gcc/9/default.nix
+++ w/pkgs/development/compilers/gcc/9/default.nix
@@ -168,7 +168,7 @@ stdenv.mkDerivation ({
     ++ optional targetPlatform.isLinux patchelf;
 
   buildInputs = [
-    gmp mpfr libmpc libelf
+    mpfr gmp libmpc libelf
     targetPackages.stdenv.cc.bintools # For linking code at run-time
   ] ++ (optional (isl != null) isl)
     ++ (optional (zlib != null) zlib)
```

It will also trigger a rebuild of the world.

The problem is that `mkDerivation` doesn't know how to filter out the unnecessary information for our `default-builder.sh` though I can't think of a better way to design it.

-----

structuredAttrs could perhaps allow us to write a `default-builder.sh` that will hopefully be resilient to this false interpretations of changed inputs.

**Comment by jtojnar** [comment]
@doronbehar order of build inputs matters since it can change the order setup hooks are run and we cannot guarantee arbitrary bash code commutes. Another example where it is a problem are environment variables like `PKG_CONFIG_PATH` – when multiple different derivations contain the same library, the first one found in the `PKG_CONFIG_PATH` will be chosen. I routinely encounter this when overriding subset of dependencies with `enableDebugging`, especially when dependency propagation comes into play.

**Review by FRidh — CHANGES_REQUESTED** [review]
This should not be merged until consensus is found in https://github.com/NixOS/nixpkgs/issues/103997.

**Comment by AndersonTorres** [comment]
This is just a parallel work for now, and way far from a clear finish.

**Comment by AndersonTorres** [comment]
Another day, another huma nerror. Now I will try to recover.

**Comment by Atemu** [comment]
GH thinks you force-pushed to a commit already in master and, since that means there are no changes, won't let you open the PR.

Should work again if you push another commit.

Would probably also not be a bad idea to simply push commits and then squash them up once they're ready. Following changes between force-pushes isn't really something you can do very well in git.

**Comment by SuperSandro2000** [comment]
> Would probably also not be a bad idea to simply push commits and then squash them up once they're ready. Following changes between force-pushes isn't really something you can do very well in git.

Not true as long as you don't do a rebase in the same force push. https://github.com/NixOS/nixpkgs/compare/32986af10085058e08125720467d84f5e1afe4b3..712c2405b1c4ddf461a1eafd640165583d1c0cca

**Comment by AndersonTorres** [comment]
Reopening
https://github.com/NixOS/nixpkgs/pull/104662

**Pull request closed without merging**